### ISSUE TITLE
$(DefaultItemExcludes) in targets contains double slashes

### ISSUE THREAD

**Issue by dasMulli** [dotnet/sdk#1375]
I just stumbled on some weird behaviour trying to add generated compile files in a target:

on macOS, having only a `Program.cs.template` file as an example (remove `Program.cs` after each build) and a leftover assembly info file in a RID specific path in `obj/` from a previous RID-specific publish:

```xml
  <Target Name="GenerateSomeFiles" BeforeTargets="BeforeBuild">
    <Exec Command="cp Program.cs.template Program.cs" />
    <Message Importance="high" Text="DefaultItemExcludes: $(DefaultItemExcludes)" />
    <Message Importance="high" Text="Exclude: $(DefaultItemExcludes);$(DefaultExcludesInProjectFolder);@(Compile)" />
    <ItemGroup>
      <Compile Include="**/*$(DefaultLanguageSourceExtension)"
               Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder);@(Compile)" />
    </ItemGroup>
    <Message Importance="high" Text="Compile: @(Compile)" />
  </Target>
```

prints:

```
$ dotnet build
Microsoft (R) Build Engine version 15.3.388.41745 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  DefaultItemExcludes: ;bin/Debug//**;obj/Debug//**;bin//**;obj//**;**/*.user;**/*.*proj;**/*.sln;**/*.vssscc;packages/**
  Exclude: ;bin/Debug//**;obj/Debug//**;bin//**;obj//**;**/*.user;**/*.*proj;**/*.sln;**/*.vssscc;packages/**;;**/.*/**;
  Compile: obj/Debug/netcoreapp2.0/osx.10.12-x64/dasmulli.sample.application.AssemblyInfo.cs;obj/Debug/netcoreapp2.0/osx.10.12-x64/sampleapp.AssemblyInfo.cs;obj/Debug/netcoreapp2.0/sampleapp.AssemblyInfo.cs;Program.cs
```

and then fails because an old assembly info file was included in `@(Compile)` (might as well have been from switching between Debug and Release configurations).

The interesting part here is that these parts have double slashes in them:
```
bin/Debug//**;obj/Debug//**;bin//**;obj//**
```

which seem to originate from [being defined like `$(BaseIntermediateOutputPath)/**`](https://github.com/dotnet/sdk/blob/30baeae0a60f26214ffad743ddbbe4851375b4a2/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets#L29-L32).

If I add `$(BaseIntermediateOutputPath)**;$(BaseOutputPath)**` to the `Exclude` attribute, the directories are excluded as expected.

However, I'd expect `$(DefaultItemExcludes)` to give me a safe value in all targets because the SDK should know which directories should be excluded and not require target authors to fix missing directories that only work during static evaluation. Is there anything that could be done to avoid this problem?

cc @dsplaisted

**Comment by nguerrera** [comment]
@dasMulli Does Exclude behave differently in static evaluation vs. targets with respect to double slashes?

**Comment by dasMulli** [comment]
Yes that was exactly the behaviour and I couldn't get my head around it. 

Context: https://stackoverflow.com/questions/44818730/is-there-a-net-core-cli-pre-before-build-task

When copying the SDK's default item definition for `Compile`, I needed to add `$(BaseIntermediateOutputPath)**;$(BaseOutputPath)**` to the `Exclude` inside a target to ensure none of any existing generated content in `obj` would be included - which should have been covered using `DefaultItemExcludes`.

**Comment by dasMulli** [comment]
@nguerrera do you think this should be fixed in msbuild only or also be changed in the targets?
cc @cdmihai who likely has knowledge about the exclude pattern implementations

**Comment by cdmihai** [comment]
@dasMulli can you please give me some repro steps? I assume the problem is that the *.cs file under obj are showing up, even though they are excluded.

I tried this on macOS and it works fine:

```xml
<Project>
  <PropertyGroup>
    <Excludes>a/b//**;c//**</Excludes>
    <Excludes>;$(Excludes);;obj/Debug//**;obj//**;</Excludes>
  </PropertyGroup>

  <ItemGroup>
    <StaticItem Include="**/*.cs" Exclude="$(Excludes)"/>
  </ItemGroup>

  <Target Name="Build">
    <ItemGroup>
      <DynamicItem Include="**/*.cs" Exclude="$(Excludes)"/>
    </ItemGroup>
    <Message Importance="high" Text="StaticItem=[@(StaticItem)]"/>
    <Message Importance="high" Text="DynamicItem=[@(DynamicItem)]"/>
  </Target>
</Project>
```

```
├── a
│   └── b
│       └── ab.cs
├── build.proj
├── c
│   └── c.cs
└── obj
   └── Debug
       └── netcoreapp2.0
           └── osx.10.12.-x64
               └── dasmulli.sample.application.AssemblyInfo.cs
```

```
~/dotnet/dotnet --version
2.2.0-preview1-007860

~/dotnet/dotnet build
Microsoft (R) Build Engine version 15.6.22.57775 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

 StaticItem=[]
 DynamicItem=[]

Build succeeded.
   0 Warning(s)
   0 Error(s)
```

**Comment by dasMulli** [comment]
Sorry I didn't try with a recent MSBuild version, this has been fixed with MSBuild 15.5 (2.1.* CLI versions,  still repros with 2.0.* CLI versions).

The issue seemed to have been a leading semicolon in the exclude patterns and not slash collapsing.

Simpler repro:

```xml
<Project DefaultTargets="CreateAndPrintSomeItems">

  <Target Name="CreateAndPrintSomeItems">
    <PropertyGroup>
      <TheExcludes>;subdir//**</TheExcludes>
    </PropertyGroup>
    <ItemGroup>
      <RuntimeItem Include="**/*.txt"
               Exclude="$(TheExcludes)" />
    </ItemGroup>
    <Message Importance="high" Text="excludes used: $(TheExcludes)" />
    <Message Importance="high" Text="RuntimeItem: %(RuntimeItem.Identity)" />
  </Target>

</Project>
```

```
├── a.txt
├── subdir
│   └── b.txt
├── test.proj
```

CLI 2.0.3:
```
Microsoft (R) Build Engine version 15.4.8.50001 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  excludes used: ;subdir//**
  RuntimeItem: a.txt
  RuntimeItem: subdir/b.txt
```

CLI 2.1.3:
```
Microsoft (R) Build Engine version 15.5.179.9764 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  excludes used: ;subdir//**
  RuntimeItem: a.txt
```
